utils/mutex-utils.js

```diff
@@ -0,0 +1,69 @@
+const AWS = require('aws-sdk');
+const { serializeError } = require('serialize-error')
+
+const logger = require('../logger/logger')
+const { GetMutexError, ReleaseMutexError } = require('../errors/MutexError')
+
+// Env vars
+const { NODE_ENV, MUTEX_TABLE_NAME } = process.env
+const IS_LOCAL_DEV = NODE_ENV === 'LOCAL_DEV'
+const mockMutexObj = {}
+
+// Dynamodb constants
+const AWS_REGION_NAME = 'ap-southeast-1'
+AWS.config.update({region: AWS_REGION_NAME})
+const docClient = new AWS.DynamoDB.DocumentClient();
+
+const mockLock = (siteName) => {
+  if (mockMutexObj[siteName]) throw new Error('Mock lock error')
+  mockMutexObj[siteName] = true
+}
+
+const mockUnlock = (siteName) => {
+  mockMutexObj[siteName] = false
+}
+
+const lock = async (siteName) => {
+  try {
+    if (!IS_LOCAL_DEV) {
+      const params = {
+        TableName: MUTEX_TABLE_NAME,
+        Item: {
+          repo_id: siteName
+        },
+        ConditionExpression: "attribute_not_exists(repo_id)"
+      }
+      await docClient.put(params).promise()  
+    } else {
+      return mockLock(siteName)
+    }
+
+    return logger.info(`Successfully locked repo ${siteName}`)
+  } catch (err) {
+    logger.error(`Failed to lock repo ${siteName}: ${JSON.stringify(serializeError(err))}`)
+    throw new GetMutexError(`Someone else is currently modifying repo ${siteName}. Please try again later.`)
+  }
+}
+
+const unlock = async (siteName) => {
+  if (IS_LOCAL_DEV) return mockUnlock(siteName)
+
+  try {
+    const params = {
+      TableName: MUTEX_TABLE_NAME,
+      Key: {
+        repo_id: siteName
+      }
+    }
+    await docClient.delete(params).promise()
+    return logger.info(`Successfully unlocked repo ${siteName}`)
+  } catch (err) {
+    logger.error(`Failed to unlock repo ${siteName}: ${JSON.stringify(serializeError(err))}`)
+    throw new ReleaseMutexError(`Something went wrong.`)
+  }
+}
```

Comment:
We could consider retrying the unlocking (using exponential backoff?), in case the unlock call fails for some reason. As discussed, we shouldn't do this for locking, since the user should just be asked to retry if the call fails.

One other possible alternative we could explore would be setting a short TTL on dynamodb, which ensures that other users can access the db after some time, even if the call to unlock fails.

Reply:
Addressed with 1dfa14c - each DynamoDB record has an `expdate` of current date + 60 seconds. However, it is important to note that DynamoDB does not delete these records immediately upon expdate.

From the [docs](https://aws.amazon.com/premiumsupport/knowledge-center/dynamodb-ttl-items-not-deleted/):
```
Wait at least for 48 hours for DynamoDB to delete the item. DynamoDB deletes expired items on a best-effort basis to be sure that there's enough throughput for other data operations. Processing takes place automatically, in the background, and doesn't affect read or write traffic to the table.
```